cdap-proto/src/main/java/io/cdap/cdap/proto/operation/OperationMeta.java

```diff
@@ -80,4 +81,66 @@ public Instant getCreateTime() {
   public Instant getEndTime() {
     return endTime;
   }
+
+  /**
+   * Creates a Builder.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Create a Builder from existing run.
+   *
+   * @param meta existing meta to copy fields from
+   */
+  public static Builder builder(OperationMeta meta) {
+    return new Builder(meta);
+  }
+
+  /**
+   * Builder to create OperationMeta.
+   */
+  @SuppressWarnings("unchecked")
+  public static class Builder {
+
+    private Set<OperationResource> resources;
+    private Instant createTime;
+    private Instant endTime;
+
+    protected Builder() {
+      this.resources = Collections.emptySet();
+    }
+
+    protected Builder(OperationMeta operationMeta) {
+      this.resources = operationMeta.resources;
+      this.createTime = operationMeta.createTime;
+      this.endTime = operationMeta.endTime;
+    }
+
+    public Builder setResources(Set<OperationResource> resources) {
+      this.resources = resources;
```

Comment:
safer to be defensive here and clear the existing resources, then add all the input (would need to change the constructor to create a modifiable Set)

Reply:
the existing code should also be returning an UnmodifiableSet for getResources()

Reply:
done